File: 1519-number-of-nodes-in-the-sub-tree-with-the-same-label/1519-number-of-nodes-in-the-sub-tree-with-the-same-label.py
class Solution:
    def countSubTrees(self, n: int, edges: List[List[int]], labels: str) -> List[int]:
        cnt = [[0 for j in range(26)] for i in range(n)]
        adj = [[] for _ in range(n)]
        for u,v in edges:
            adj[u].append(v)
            adj[v].append(u)
        l = [ord(labels[i]) - ord('a') for i in range(n)]
        
        def dfs(u, p, c):
            cnt[u][c] += (c==l[u])
            for v in adj[u]:
                if v == p: continue
                cnt[u][c] += dfs(v,u,c)
            return cnt[u][c]
        
        for c in range(26): dfs(0,-1,c)
            
        ret = [cnt[i][l[i]] for i in range(n)]
        return ret

File: 1519-number-of-nodes-in-the-sub-tree-with-the-same-label/1519-number-of-nodes-in-the-sub-tree-with-the-same-label.cpp
class Solution {
public:
    
    int n;
    vector<vector<int>> cnt;
    vector<vector<int>> adj;
    vector<int> l;
    
    int dfs(int u, int p, int c) {
        cnt[u][c] += (c==l[u]);
        for (auto v : adj[u]) {
            if (v == p) continue;
            cnt[u][c] += dfs(v, u, c);            
        }
        return cnt[u][c];
    }
    
    vector<int> countSubTrees(int N, vector<vector<int>>& edges, string labels) {
        n = N;
        vector<vector<int>> temp_cnt(n, vector<int>(26));  // all initialized with 0
        cnt = temp_cnt;
        adj.resize(n);
        for (auto uv : edges) {
            int u = uv[0];
            int v = uv[1];
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        l.resize(n);
        for (int i=0; i<n; i++) l[i] = labels[i] - 'a';
        
        for (int c=0; c<26; c++) {
            dfs(0,-1,c);
        }
        vector<int> ret(n);
        for (int i=0; i<n; i++) {
            ret[i] = cnt[i][l[i]];
        }
        
        
        return ret;
        
    }
};